File: ios/SiriApp/Config/ConfigViewModel.swift
//
//  ConfigViewModel.swift
//  SiriApp
//
//  Settings ViewModel: save / test / clear config.
//  Ported from Android: ConfigViewModel.kt
//

import Foundation
import Combine

enum ConnectionTestState: Equatable {
    case idle
    case testing
    case success(String)
    case failure(String)
}

class ConfigViewModel: ObservableObject {
    @Published var config: LlmConfig?
    @Published var testResult: ConnectionTestState = .idle

    private let repository = ConfigRepository()

    init() {
        config = repository.getConfig()
    }

    func saveConfig(_ apiUrl: String, _ model: String, _ apiKey: String) {
        let trimmedUrl = apiUrl.trimmingCharacters(in: .whitespaces)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let trimmedModel = model.trimmingCharacters(in: .whitespaces)
        let trimmedKey = apiKey.trimmingCharacters(in: .whitespaces)

        guard !trimmedUrl.isEmpty, !trimmedModel.isEmpty, !trimmedKey.isEmpty else {
            testResult = .failure("所有字段不能为空")
            return
        }

        guard trimmedUrl.hasPrefix("http://") || trimmedUrl.hasPrefix("https://") else {
            testResult = .failure("API 地址必须以 http:// 或 https:// 开头")
            return
        }

        let newConfig = LlmConfig(
            apiUrl: trimmedUrl,
            model: trimmedModel,
            apiKey: trimmedKey
        )
        repository.saveConfig(newConfig)
        config = newConfig
        testResult = .success("配置已保存")
    }

    func testConnection(_ apiUrl: String, _ model: String, _ apiKey: String) {
        let trimmedUrl = apiUrl.trimmingCharacters(in: .whitespaces)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let trimmedModel = model.trimmingCharacters(in: .whitespaces)
        let trimmedKey = apiKey.trimmingCharacters(in: .whitespaces)

        guard !trimmedUrl.isEmpty, !trimmedModel.isEmpty, !trimmedKey.isEmpty else {
            testResult = .failure("请先填写完整信息")
            return
        }

        testResult = .testing

        // Test both LLM and Embedding APIs
        Task {
            let llmResult = await testLLM(baseUrl: trimmedUrl, model: trimmedModel, apiKey: trimmedKey)
            let embResult = await testEmbedding(baseUrl: trimmedUrl, apiKey: trimmedKey)

            await MainActor.run {
                switch (llmResult, embResult) {
                case (.success, .success):
                    testResult = .success("连接成功！LLM OK, Embedding OK")
                case (.success, .failure(let embErr)):
                    testResult = .failure("LLM OK, 但 Embedding 失败: \(embErr)")
                case (.failure(let llmErr), .success):
                    testResult = .failure("LLM 失败: \(llmErr)")
                case (.failure(let llmErr), .failure(let embErr)):
                    testResult = .failure("LLM 失败: \(llmErr)\nEmbedding 也失败: \(embErr)")
                }
            }
        }
    }

    // MARK: - Individual API tests

    private enum ApiTestResult {
        case success
        case failure(String)
    }

    private func testLLM(baseUrl: String, model: String, apiKey: String) async -> ApiTestResult {
        guard let url = URL(string: "\(baseUrl)/chat/completions") else {
            return .failure("无效的 API 地址")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 10

        let body: [String: Any] = [
            "model": model,
            "messages": [["role": "user", "content": "hi"]],
            "max_tokens": 1,
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        return await withCheckedContinuation { continuation in
            URLSession.shared.dataTask(with: request) { data, response, error in
                if let error = error {
                    continuation.resume(returning: .failure(self.summarizeError(error)))
                    return
                }
                guard let httpResponse = response as? HTTPURLResponse else {
                    continuation.resume(returning: .failure("无效的服务器响应"))
                    return
                }
                if (200...299).contains(httpResponse.statusCode) {
                    continuation.resume(returning: .success)
                } else {
                    let errorBody = data.flatMap { String(data: $0, encoding: .utf8) } ?? "未知错误"
                    continuation.resume(returning: .failure("\(httpResponse.statusCode) \(String(errorBody.prefix(100)))"))
                }
            }.resume()
        }
    }

    private func testEmbedding(baseUrl: String, apiKey: String) async -> ApiTestResult {
        guard let url = URL(string: "\(baseUrl)/embeddings") else {
            return .failure("无效的 API 地址")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 10

        let embeddingModel = repository.getEmbeddingModel()
        let body: [String: Any] = [
            "model": embeddingModel,
            "input": "test",
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        return await withCheckedContinuation { continuation in
            URLSession.shared.dataTask(with: request) { data, response, error in
                if let error = error {
                    continuation.resume(returning: .failure(self.summarizeError(error)))
                    return
                }
                guard let httpResponse = response as? HTTPURLResponse else {
                    continuation.resume(returning: .failure("无效的服务器响应"))
                    return
                }
                if (200...299).contains(httpResponse.statusCode) {
                    continuation.resume(returning: .success)
                } else {
                    let errorBody = data.flatMap { String(data: $0, encoding: .utf8) } ?? "未知错误"
                    continuation.resume(returning: .failure("\(httpResponse.statusCode) \(String(errorBody.prefix(100)))"))
                }
            }.resume()
        }
    }

    private func summarizeError(_ error: Error) -> String {
        let nsError = error as NSError
        switch nsError.code {
        case NSURLErrorCannotFindHost, NSURLErrorCannotConnectToHost:
            return "无法连接到服务器，请检查 API 地址"
        case NSURLErrorTimedOut:
            return "连接超时，请检查网络"
        case NSURLErrorServerCertificateUntrusted:
            return "SSL 证书验证失败"
        default:
            return error.localizedDescription
        }
    }

    func clearConfig() {
        repository.clearConfig()
        config = nil
        testResult = .idle
    }

    func resetTestResult() {
        testResult = .idle
    }
}

File: android/app/src/main/java/com/rd/siri/config/ConfigViewModel.kt
package com.rd.siri.config

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

sealed class ConnectionTestResult {
    object Idle : ConnectionTestResult()
    object Testing : ConnectionTestResult()
    data class Success(val message: String) : ConnectionTestResult()
    data class Failure(val error: String) : ConnectionTestResult()
}

class ConfigViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = ConfigRepository(application)

    private val _config = MutableStateFlow(repository.getConfig())
    val config: StateFlow<LlmConfig?> = _config.asStateFlow()

    private val _testResult = MutableStateFlow<ConnectionTestResult>(ConnectionTestResult.Idle)
    val testResult: StateFlow<ConnectionTestResult> = _testResult.asStateFlow()

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    fun saveConfig(apiUrl: String, model: String, apiKey: String, enableSearch: Boolean = false, searchParamName: String = "enable_search") {
        val trimmedUrl = apiUrl.trim().trimEnd('/')
        val trimmedModel = model.trim()
        val trimmedKey = apiKey.trim()

        if (trimmedUrl.isBlank() || trimmedModel.isBlank() || trimmedKey.isBlank()) {
            _testResult.value = ConnectionTestResult.Failure("所有字段不能为空")
            return
        }

        if (!trimmedUrl.startsWith("http://") && !trimmedUrl.startsWith("https://")) {
            _testResult.value = ConnectionTestResult.Failure("API 地址必须以 http:// 或 https:// 开头")
            return
        }

        val newConfig = LlmConfig(trimmedUrl, trimmedModel, trimmedKey, enableSearch, searchParamName)
        repository.saveConfig(newConfig)
        _config.value = newConfig
        _testResult.value = ConnectionTestResult.Success("配置已保存")
    }

    fun testConnection(apiUrl: String, model: String, apiKey: String) {
        if (apiUrl.isBlank() || model.isBlank() || apiKey.isBlank()) {
            _testResult.value = ConnectionTestResult.Failure("请先填写完整信息")
            return
        }

        viewModelScope.launch {
            _testResult.value = ConnectionTestResult.Testing

            val baseUrl = apiUrl.trim().trimEnd('/')
            val key = apiKey.trim()
            val llmModel = model.trim()

            val llmResult = withContext(Dispatchers.IO) { testLLM(baseUrl, llmModel, key) }
            val embResult = withContext(Dispatchers.IO) { testEmbedding(baseUrl, key) }

            _testResult.value = when {
                llmResult is ConnectionTestResult.Success && embResult is ConnectionTestResult.Success ->
                    ConnectionTestResult.Success("连接成功！LLM OK, Embedding OK")
                llmResult is ConnectionTestResult.Success && embResult is ConnectionTestResult.Failure ->
                    ConnectionTestResult.Failure("LLM OK, 但 Embedding 失败: ${embResult.error}")
                llmResult is ConnectionTestResult.Failure && embResult is ConnectionTestResult.Success ->
                    ConnectionTestResult.Failure("LLM 失败: ${llmResult.error}")
                llmResult is ConnectionTestResult.Failure && embResult is ConnectionTestResult.Failure ->
                    ConnectionTestResult.Failure("LLM 失败: ${llmResult.error}\nEmbedding 也失败: ${embResult.error}")
                else -> ConnectionTestResult.Failure("未知错误")
            }
        }
    }

    private fun testLLM(baseUrl: String, model: String, apiKey: String): ConnectionTestResult {
        return try {
            val body = JSONObject().apply {
                put("model", model)
                put("messages", JSONArray().apply {
                    put(JSONObject().apply {
                        put("role", "user")
                        put("content", "hi")
                    })
                })
                put("max_tokens", 1)
            }

            val request = Request.Builder()
                .url("$baseUrl/chat/completions")
                .addHeader("Authorization", "Bearer $apiKey")
                .addHeader("Content-Type", "application/json")
                .post(body.toString().toRequestBody("application/json".toMediaType()))
                .build()

            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                ConnectionTestResult.Success("OK")
            } else {
                val errorBody = response.body?.string() ?: "未知错误"
                ConnectionTestResult.Failure("${response.code} ${errorBody.take(100)}")
            }
        } catch (e: java.net.UnknownHostException) {
            ConnectionTestResult.Failure("无法解析地址")
        } catch (e: java.net.ConnectException) {
            ConnectionTestResult.Failure("连接被拒绝")
        } catch (e: java.net.SocketTimeoutException) {
            ConnectionTestResult.Failure("连接超时")
        } catch (e: javax.net.ssl.SSLHandshakeException) {
            ConnectionTestResult.Failure("SSL 证书验证失败")
        } catch (e: Exception) {
            ConnectionTestResult.Failure(e.message ?: "未知错误")
        }
    }

    private fun testEmbedding(baseUrl: String, apiKey: String): ConnectionTestResult {
        return try {
            val embeddingModel = repository.getEmbeddingModel()
            val body = JSONObject().apply {
                put("model", embeddingModel)
                put("input", "test")
            }

            val request = Request.Builder()
                .url("$baseUrl/embeddings")
                .addHeader("Authorization", "Bearer $apiKey")
                .addHeader("Content-Type", "application/json")
                .post(body.toString().toRequestBody("application/json".toMediaType()))
                .build()

            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                ConnectionTestResult.Success("OK")
            } else {
                val errorBody = response.body?.string() ?: "未知错误"
                ConnectionTestResult.Failure("${response.code} ${errorBody.take(100)}")
            }
        } catch (e: java.net.UnknownHostException) {
            ConnectionTestResult.Failure("无法解析地址")
        } catch (e: java.net.ConnectException) {
            ConnectionTestResult.Failure("连接被拒绝")
        } catch (e: java.net.SocketTimeoutException) {
            ConnectionTestResult.Failure("连接超时")
        } catch (e: javax.net.ssl.SSLHandshakeException) {
            ConnectionTestResult.Failure("SSL 证书验证失败")
        } catch (e: Exception) {
            ConnectionTestResult.Failure(e.message ?: "未知错误")
        }
    }

    fun clearConfig() {
        repository.clearConfig()
        _config.value = null
        _testResult.value = ConnectionTestResult.Idle
    }

    fun resetTestResult() {
        _testResult.value = ConnectionTestResult.Idle
    }
}
